### DISCUSSION TITLE
Can't Get Single Element Properties when using getElementProperties api

### DISCUSSION THREAD

**Discussion by ghost** [iTwin/itwinjs-core#6619] [General] [answered]
I wrote a test class, but when calling the PresentationManager.getElementProperties function, I couldn't retrieve the element's property information. I'm not sure where I went wrong, thank you for your help. Below is the code for the test class and the call:

**// Test Class**
```
export class PropertiesService {
  private _imodelfile: string;
  private _presentationManager: PresentationManager;
  private _imodel: IModelDb;
  private _elementClasses: string[];

  constructor(imodelfile: string, elementClasses: string[]) {
    this._imodelfile = imodelfile;
    this._elementClasses = elementClasses;
    this.init();
  }

  async retrieveElementProperties(elementId: string) {
    console.log(this._imodelfile);
    this._imodel = await SnapshotDb.openFile(this._imodelfile);
    const options: SingleElementPropertiesRequestOptions<IModelDb> = {
      imodel: this._imodel,
      elementId
    };
    const properties = await this._presentationManager.getElementProperties(options);
    return { elementId, properties };
  }

  private init() {
    Presentation.initialize({
      requestTimeout: 2000 * 60 * 1000,
      useSingleManager: true,
      caching: {
        content: { size: 10000 },
        // workerConnectionCacheSize: 32768000 * 100,
        hierarchies: { mode: HierarchyCacheMode.Memory }
      }
    });
    this._presentationManager = Presentation.getManager();
    this._presentationManager.rulesets().clear();
  }
}
```

**// Test Code**
```
  async retrieveElementProperties(ctx, _next) {
    const imodelGuidKeyName = Object.keys(ctx.params)[0];
    const elementIdName = Object.keys(ctx.params)[1];
    const imodelGuid = ctx.params[imodelGuidKeyName];
    const elementId = ctx.params[elementIdName];

    if (!imodelGuid || !elementId) {
      ctx.body = 'modelGuid、elementId不能为空';
    }
    const filepath = path.normalize(path.join(OUTPUT_DIR, imodelGuid, '0', `${imodelGuid}.bim`));
    const elementClasses = ['BisCore.GeometricElement2D', 'BisCore.GeometricElement3D'];
    const propertiesService = new PropertiesService(filepath, elementClasses);

    try {
      const propertyCategoriesData = await propertiesService.retrieveElementProperties(elementId);
      ctx.body = propertyCategoriesData;
    } catch (error) {
      propertiesService.dispose();
      ctx.body = `Retrieve Element's Properties error`;
    }
  }
```

![property](https://github.com/iTwin/itwinjs-core/assets/34014227/c3362d05-21a6-4051-8288-29d1b07d38a2)

**Comment by grigasp** [comment]
Hi, can you please be more specific what you mean by "couldn't retrieve the element's property information"? 
- Are you getting an error? If yes, please show us the error message and a call stack, if any.
- Is the result different from what you expect? If yes, please explain what you expect and show what you get.

Thanks.

**Reply by ghost** [reply]
After calling this function, the returned 'properties' is undefined, and there is no error message provided.
```
const properties = await this._presentationManager.getElementProperties(options);
```
I reviewed the source code of PresentationManager.ts file:
https://github.com/iTwin/itwinjs-core/blob/dbab3d1837bc170fdcd2b72330288400341b8c6e/presentation/backend/src/presentation-backend/PresentationManager.ts
Inside, `getElementProperties` actually invokes `getContent`. When I directly called `getContent`, the returned 'content' result was also undefined.

```
async retrieveElementProperties(elementId: string) {
    console.log(this._imodelfile);
    this._imodel = await SnapshotDb.openFile(this._imodelfile);
    const options: SingleElementPropertiesRequestOptions<IModelDb> = {
      imodel: this._imodel,
      elementId
    };

    let properties: ElementProperties[] = [];
    if (isSingleElementPropertiesRequestOptions(options)) {
      const { elementId, ...optionsNoElementId } = options;
      const content = await this._presentationManager.getContent({
        ...optionsNoElementId,
        descriptor: {
          displayType: DefaultContentDisplayTypes.PropertyPane,
          contentFlags: ContentFlags.ShowLabels
        },
        rulesetOrId: 'ElementProperties',
        keys: new KeySet([{ className: 'BisCore:Element', id: elementId }])
      });
      console.log(content); // **!!! the content is undefined !!!**
      properties = buildElementsProperties(content);
    }

    return { elementId, properties };
  }
```
This is the bim file I tested with, and the elemntId I passed in is "0x153".
[57a1bc88-e624-4f13-94d8-547fb0cf6a3f.zip](https://github.com/iTwin/itwinjs-core/files/14974844/57a1bc88-e624-4f13-94d8-547fb0cf6a3f.zip)

Thanks for your help.

**Reply by grigasp** [reply]
That most likely means that `elementId` is invalid (not a valid id, element with the id doesn't exist or something like that). Can you confirm validity of that id?

**Reply by ghost** [reply]
@grigasp  the elemntId I passed in is "0x153", and comfirmed that it is a valid id. see the bim file i uploaded above. Here's a screenshot:
![01](https://github.com/iTwin/itwinjs-core/assets/34014227/ec6f072b-7b83-4a35-b6fd-279218052e6c)
![02](https://github.com/iTwin/itwinjs-core/assets/34014227/3e340368-53bd-45c5-acf9-b8fdf3b1dca8)

**Reply by grigasp** [reply]
I can't reproduce this on `master` (see the test [here](https://github.com/iTwin/itwinjs-core/compare/release/4.5.1...grigasp/discussion-6619)). What version of `itwinjs-core` are you using?

**Reply by ghost** [reply]
@grigasp my package.json's dependencies:
```
  "dependencies": {
    "@itwin/core-backend": "4.5.1",
    "@itwin/core-bentley": "4.5.1",
    "@itwin/core-common": "4.5.1",
    "@itwin/presentation-backend": "^4.5.1",
    "@itwin/presentation-common": "^4.5.1",
    "@koa/cors": "^5.0.0",
    "@koa/multer": "^3.0.2",
    "@koa/router": "^12.0.1",
    "dotenv": "^16.3.1",
    "koa": "^2.14.2",
    "koa-bodyparser": "^4.4.1",
    "koa-mount": "^4.0.0",
    "koa-static": "^5.0.0",
    "rimraf": "^3.0.2",
    "uuid": "^9.0.1"
  },
  "devDependencies": {
    "@types/jsonwebtoken": "^9.0.5",
    "@types/koa__multer": "^2.0.7",
    "@types/koa__router": "^12.0.4",
    "@types/koa-bodyparser": "^4.3.12",
    "@types/node": "^20.10.6",
    "@types/rimraf": "^3.0.0",
    "concurrently": "^8.2.2",
    "nodemon": "^3.0.2"
  }
```

**Reply by grigasp** [reply, marked as answer]
I rebased [my branch](https://github.com/iTwin/itwinjs-core/compare/release/4.5.1...grigasp/discussion-6619) on `release/4.5.1` and the test still succeeds. 

Now that I look at your initial post, I think the problem is this line:
```ts
this._presentationManager.rulesets().clear();
```
When a `PresentationManager` is created, it registers the rulesets it uses, including the "ElementProperties" one. Your call clears the registered rulesets list, so requesting content with "ElementProperties" ruleset ID results in no content.

**Reply by ghost** [reply]
Thank you, just as you said! But I'm also sorry for taking up so much of your time.
